File: engine/dynamic/render_vulkan/src/util/command_buffer.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/logging.hpp"

#include "internal/render_vulkan/module_render_vulkan.hpp"
#include "internal/render_vulkan/util/command_buffer.hpp"
#include "internal/render_vulkan/state/renderer_state.hpp"

#include "vulkan/vulkan.h"

#include <algorithm>

namespace argus {
    VkCommandPool create_command_pool(const LogicalDevice &device, uint32_t queue_index) {
        VkCommandPoolCreateInfo pool_info {};
        pool_info.sType = VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO;
        pool_info.flags = VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT;
        pool_info.queueFamilyIndex = queue_index;

        VkCommandPool command_pool {};
        if (vkCreateCommandPool(device.logical_device, &pool_info, nullptr, &command_pool) != VK_SUCCESS) {
            Logger::default_logger().fatal("Failed to create command pool");
        }

        return command_pool;
    }

    void destroy_command_pool(const LogicalDevice &device, VkCommandPool command_pool) {
        vkDestroyCommandPool(device.logical_device, command_pool, nullptr);
    }

    std::vector<CommandBufferInfo> alloc_command_buffers(const LogicalDevice &device, VkCommandPool pool,
            uint32_t count) {
        VkCommandBufferAllocateInfo cb_alloc_info {};
        cb_alloc_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
        cb_alloc_info.commandPool = pool;
        cb_alloc_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
        cb_alloc_info.commandBufferCount = count;

        std::vector<VkCommandBuffer> handles;
        handles.resize(count);
        if (vkAllocateCommandBuffers(device.logical_device, &cb_alloc_info, handles.data()) != VK_SUCCESS) {
            Logger::default_logger().fatal("Failed to allocate command buffers");
        }

        std::vector<CommandBufferInfo> buffers;
        buffers.resize(count);
        std::transform(handles.cbegin(), handles.cend(), buffers.begin(), [&pool](auto handle) {
            CommandBufferInfo cb_info {};
            cb_info.pool = pool;
            cb_info.handle = handle;
            return cb_info;
        });

        return buffers;
    }

    void free_command_buffers(const LogicalDevice &device, const std::vector<CommandBufferInfo> &buffers) {
        if (buffers.empty()) {
            return;
        }

        std::vector<VkCommandBuffer> handles;
        handles.reserve(buffers.size());
        std::transform(buffers.cbegin(), buffers.cend(), handles.begin(), [](const auto &buf) { return buf.handle; });

        vkFreeCommandBuffers(device.logical_device, buffers.front().pool,
                static_cast<uint32_t>(handles.size()), handles.data());
    }

    void free_command_buffer(const LogicalDevice &device, const CommandBufferInfo &buffer) {
        vkFreeCommandBuffers(device.logical_device, buffer.pool, 1, &buffer.handle);
    }

    void begin_oneshot_commands(const LogicalDevice &device, const CommandBufferInfo &buffer) {
        UNUSED(device);

        vkResetCommandBuffer(buffer.handle, 0);

        VkCommandBufferBeginInfo begin_info {};
        begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
        begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
        vkBeginCommandBuffer(buffer.handle, &begin_info);
    }

    void end_command_buffer(const LogicalDevice &device, const CommandBufferInfo &buffer) {
        UNUSED(device);

        vkEndCommandBuffer(buffer.handle);
    }

    void submit_command_buffer(const LogicalDevice &device, const CommandBufferInfo &buffer, VkQueue queue,
            VkFence fence, const std::vector<VkSemaphore> &wait_semaphores,
            const std::vector<VkPipelineStageFlags> &wait_stages,
            const std::vector<VkSemaphore> &signal_semaphores) {
        UNUSED(device);

        assert(wait_semaphores.size() == wait_stages.size());

        VkSubmitInfo submit_info {};
        submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
        submit_info.commandBufferCount = 1;
        submit_info.pCommandBuffers = &buffer.handle;
        submit_info.waitSemaphoreCount = uint32_t(wait_semaphores.size());
        submit_info.pWaitSemaphores = wait_semaphores.data();
        submit_info.pWaitDstStageMask = wait_stages.data();
        submit_info.signalSemaphoreCount = uint32_t(signal_semaphores.size());
        submit_info.pSignalSemaphores = signal_semaphores.data();
        if (vkQueueSubmit(queue, 1, &submit_info, fence) != VK_SUCCESS) {
            Logger::default_logger().fatal("Failed to submit command queues");
        }
    }

    void queue_command_buffer_submit(RendererState &state, const CommandBufferInfo &buffer, VkQueue queue,
            VkFence fence, std::vector<VkSemaphore> wait_semaphores, std::vector<VkPipelineStageFlags> wait_stages,
            std::vector<VkSemaphore> signal_semaphores, Semaphore *submit_semaphore) {
        {
            std::lock_guard<std::mutex> lock(state.submit_mutex);
            state.submit_bufs.push_back(CommandBufferSubmitParams { false, 0, state.cur_frame, &buffer, queue, fence,
                    std::move(wait_semaphores), std::move(wait_stages), std::move(signal_semaphores),
                    submit_semaphore });
        }
        state.queued_submit_sem.notify();
    }
}

File: engine/modules/render_vulkan/src/util/command_buffer.rs
use crate::setup::device::VulkanDevice;
use crate::state::{SubmitCommandBufferParams, SubmitMessage};
use argus_util::semaphore::Semaphore;
use ash::vk;
use std::sync::{mpsc, Mutex};
use std::sync::mpsc::SendError;

#[derive(Clone)]
pub(crate) struct CommandBufferInfo {
    handle: vk::CommandBuffer,
    pool: vk::CommandPool,
}

impl CommandBufferInfo {
    /// SAFETY: The returned handle must not outlive `self`.
    pub(crate) unsafe fn get_handle(&self) -> vk::CommandBuffer {
        self.handle
    }
}

pub(crate) fn create_command_pool(device: &VulkanDevice, queue_index: u32) -> vk::CommandPool {
    let pool_info = vk::CommandPoolCreateInfo::default()
        .flags(vk::CommandPoolCreateFlags::RESET_COMMAND_BUFFER)
        .queue_family_index(queue_index);

    let command_pool = unsafe {
        device.logical_device.create_command_pool(&pool_info, None)
            .expect("Failed to create command pool")
    };

    command_pool
}

pub(crate) fn destroy_command_pool(device: &VulkanDevice, command_pool: vk::CommandPool) {
    unsafe { device.logical_device.destroy_command_pool(command_pool, None) };
}

pub(crate) fn alloc_command_buffers(
    device: &VulkanDevice,
    pool: vk::CommandPool,
    count: u32,
) -> Vec<CommandBufferInfo> {
    let cb_alloc_info = vk::CommandBufferAllocateInfo::default()
        .command_pool(pool)
        .level(vk::CommandBufferLevel::PRIMARY)
        .command_buffer_count(count);

    let handles = unsafe {
        device.logical_device.allocate_command_buffers(&cb_alloc_info)
            .expect("Failed to allocate command buffers")
    };

    let buffers = handles.into_iter()
        .map(|handle| CommandBufferInfo {
            handle,
            pool,
        })
        .collect();
    buffers
}

pub(crate) fn free_command_buffers(device: &VulkanDevice, buffers: Vec<CommandBufferInfo>) {
    if buffers.is_empty() {
        return;
    }

    let handles: Vec<_> = buffers.iter().map(|buffer| buffer.handle).collect();
    unsafe { device.logical_device.free_command_buffers(buffers.first().unwrap().pool, &handles) };
}

pub(crate) fn destroy_command_buffer(device: &VulkanDevice, buffer: CommandBufferInfo) {
    unsafe { device.logical_device.free_command_buffers(buffer.pool, &[buffer.handle]); }
}

pub(crate) fn begin_oneshot_commands(device: &VulkanDevice, buffer: &CommandBufferInfo) {
    unsafe {
        device.logical_device
            .reset_command_buffer(buffer.handle, vk::CommandBufferResetFlags::empty())
            .expect("Failed to reset command buffer");
    }

    let begin_info = vk::CommandBufferBeginInfo::default()
        .flags(vk::CommandBufferUsageFlags::ONE_TIME_SUBMIT);
    unsafe {
        device.logical_device.begin_command_buffer(buffer.handle, &begin_info)
            .expect("Failed to begin command buffer"); //TODO
    };
}

pub(crate) fn end_command_buffer(device: &VulkanDevice, buffer: &CommandBufferInfo) {
    unsafe {
        device.logical_device.end_command_buffer(buffer.handle)
            .expect("Failed to end command buffer"); //TODO
    }
}

pub(crate) fn submit_command_buffer(
    device: &ash::Device,
    buffer: &CommandBufferInfo,
    queue: vk::Queue,
    fence: vk::Fence,
    wait_semaphores: Vec<vk::Semaphore>,
    wait_stages: Vec<vk::PipelineStageFlags>,
    signal_semaphores: Vec<vk::Semaphore>,
) {
    assert_eq!(wait_semaphores.len(), wait_stages.len());
    
    let buffers = [buffer.handle];

    let submit_info = vk::SubmitInfo::default()
        .command_buffers(&buffers)
        .wait_semaphores(&wait_semaphores)
        .wait_dst_stage_mask(&wait_stages)
        .signal_semaphores(&signal_semaphores);
    let submit_infos = [submit_info];
    unsafe {
        device.queue_submit(queue, &submit_infos, fence)
            .expect("Failed to submit command queues");
    }
}

pub(crate) fn queue_command_buffer_submit(
    sender: &mpsc::Sender<SubmitMessage>,
    submit_mutex: &Mutex<()>,
    buffer: CommandBufferInfo,
    queue: vk::Queue,
    wait_semaphores: Vec<vk::Semaphore>,
    wait_stages: Vec<vk::PipelineStageFlags>,
    signal_semaphores: Vec<vk::Semaphore>,
    fence: Option<vk::Fence>,
    in_flight_sem: Option<Semaphore>,
) -> Result<(), SendError<SubmitMessage>> {
    let _lock = submit_mutex.lock().unwrap();
    let params = {
        SubmitCommandBufferParams {
            buffer,
            queue,
            wait_sems: wait_semaphores,
            wait_stages,
            signal_sems: signal_semaphores,
            fence,
            in_flight_sem,
        }
    };
    sender.send(SubmitMessage::SubmitCommandBuffer(params))?;
    Ok(())
}
